Исходные данные:
1)Набор данных - https://www.kaggle.com/new-york-city/ny-2015-street-tree-census-tree-data?select=2015-street-tree-census-tree-data.csv

Задание:
1 Сколько в датасете объектов и признаков? Дать описание каждому признаку, если оно есть.

2 Сколько категориальных признаков, какие?

3 Столбец с макимальным количеством уникальных значений категориального признака?

4 Есть ли бинарные признаки?

5 Какие числовые признаки?

6 Есть ли пропуски?

7 Сколько объектов с пропусками?

8 Столбец с максимальным количеством пропусков?

9 Есть ли на ваш взгляд выбросы, аномальные значения?

10 Столбец с максимальным средним значением после нормировки признаков через стандартное отклонение?

11 Столбец с целевым признаком?

12 Сколько объектов попадает в тренировочную выборку при использовании train_test_split с параметрами test_size = 0.3, random_state = 42?

13 Между какими признаками наблюдается линейная зависимость (корреляция)?

14 Сколько признаков достаточно для объяснения 90% дисперсии после применения метода PCA?

15 Какой признак вносит наибольший вклад в первую компоненту?

Доп. задание: построить двухмерное представление данных с помощью алгоритма t-SNE. На сколько кластеров визуально на ваш взгляд разделяется выборка?

Установка необходимых библиотек:

In [ ]:
!pip install pandas
!pip install sklearn

In [18]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
data = pd.read_csv('2015-street-tree-census-tree-data.csv')

In [19]:
data.head()

tree_id  block_id               created_at  tree_dbh  stump_diam curb_loc  \
0   180683    348711  2015-08-27T00:00:00.000         3           0   OnCurb   
1   200540    315986  2015-09-03T00:00:00.000        21           0   OnCurb   
2   204026    218365  2015-09-05T00:00:00.000         3           0   OnCurb   
3   204337    217969  2015-09-05T00:00:00.000        10           0   OnCurb   
4   189565    223043  2015-08-30T00:00:00.000        21           0   OnCurb   

  status health                           spc_latin       spc_common  ...  \
0  Alive   Fair                         Acer rubrum        red maple  ...   
1  Alive   Fair                   Quercus palustris          pin oak  ...   
2  Alive   Good  Gleditsia triacanthos var. inermis      honeylocust  ...   
3  Alive   Good  Gleditsia triacanthos var. inermis      honeylocust  ...   
4  Alive   Good                     Tilia americana  American linden  ...   

   boro_ct     state   latitude  longitude         x_sp         y_sp  \
0  4073900  New York  40.723092 -73.844215  1027431.148  202756.7687   
1  4097300  New York  40.794111 -73.818679  1034455.701  228644.8374   
2  3044900  New York  40.717581 -73.936608  1001822.831  200716.8913   
3  3044900  New York  40.713537 -73.934456  1002420.358  199244.2531   
4  3016500  New York  40.666778 -73.975979   990913.775  182202.4260   

  council district census tract        bin           bbl  
0             29.0        739.0  4052307.0  4.022210e+09  
1             19.0        973.0  4101931.0  4.044750e+09  
2             34.0        449.0  3338310.0  3.028870e+09  
3             34.0        449.0  3338342.0  3.029250e+09  
4             39.0        165.0  3025654.0  3.010850e+09  

[5 rows x 45 columns]

In [20]:
data['boroname'].value_counts()

KeyError: 'boroname'

1)tree_id - уникальный идентификационный номер дерева (целое число)

2)block_id - Идентификационный номер блок-фасада, на который отображено дерево (целое число)

3)created_at - Дата нанесения деревьев на карту (дата)

4)tree_dbh - Диаметр на высоте груди дерева (целое число)

5)stump_diam - Диаметр пня (целое число)

6)curb_loc - Расположено ли дерево вдоль бордюра или со смещением от него (Текст Значения домена: OnCurb OffsetFromCurb)

7)status - Статус дерева (Текст Доменные значения: Живой, Пень, Мертвый)

8)health - Здоровье деревьев (Текст Доменные ценности: Хорошо, Неплохо, Плохо)

9)spc_latin - Научное / латинское название вида дерева (Текст)

10)spc_common - Общее название породы деревьев (Текст)

11)steward - Количество замеченных признаков управления (Текст Значения доменов: 1 или 2 ,3 или 4, 4 или больше, Ничего)

12)guards - Наличие и тип охраны деревьевНаличие и тип охраны деревьев (Текст Доменные ценности: Вредные Полезно Нет Не уверен)

13)sidewalk - Повреждение тротуара в непосредственной близости от дерева (Текст Значения домена: Повреждение, Не повреждено)

14)user_type - Категория пользователя, который собрал эту точку дерева (Текст Доменные ценности: Волонтер, TreesCount Персонал, Персонал NYC Parks)

15)root_stone - Проблемы с корнями, вызванные брусчаткой в ложе дерева (Текст Значения домена: Да Нет)

16)root_grate - Проблемы с корнями, вызванные металлическими решетками (Текст Значения домена: Да Нет)

17)root_other - Наличие других проблем с корнями (Текст Значения домена: Да Нет)

18)trunk_wire - Проблемы со стволом, вызванные тросом или проводами (Текст Значения домена: Да Нет)

19)trnk_light - Проблемы с багажником, вызванные освещением (Текст Значения домена: Да Нет)

20)trnk_other - Наличие других проблем со стволом (Текст Значения домена: Да Нет)

21)brch_light - Проблемы с ветками, вызванные освещением или проводами (Текст Значения домена: Да Нет)

22)brch_shoe - Проблемы с ветвями, вызванные обувью (Текст Значения домена: Да Нет)

23)brch_other - Наличие других проблем с ветвями (Текст Значения домена: Да Нет)

24)address - Предполагаемый уличный адрес дерева (Текст)

25)zipcode - Почтовый индекс (целое число)

26)zip_city - Город, полученный из почтового индекса (Текст)

27)cb_num - Общественный совет (Целое число)

28)borocode - Геокодирование на основе местоположения точки дерева (Целое число Доменные значения: 1 (Манхэттен) 2 (Бронкс) 3 (Бруклин) 4 (Квинс) 5 (Статен-Айленд))

29)boroname - Геокодирование на основе местоположения точки дерева (Текст Доменные ценности: Манхэттен Бронкс, Бруклин Квинс, Статен-Айленд) ?(выдаёт ошибку)

30)cncldist - Округ совета Нью-Йорка (Целое число)

31)st_assem - Округ Ассамблеи штата Нью-Йорк (Целое число)

32)st_senate - Сенатский округ штата Нью-Йорк (Целое число)

33)nta - Код района табуляции соседей (Текст)

34)nta_name - Название зоны табуляции соседей (Текст)

35)boro_ct - Переписной участок (Текст)

36)state - Государство (Текст)

37)latitude - Широта точки дерева (Двойной)

38)longitude - Долгота точки дерева (Двойной)

39)x_sp - X координата точки дерева (Двойной)

40)y_sp - Y координата точки дерева (Двойной)

ВОПРОС 2. Сколько категориальных признаков, какие?

Всего 9 категориальных признаков
1)status Статус дерева
2)health Здоровье деревьев
3)steward Количество замеченных признаков управления
4)guards Наличие и тип охраны деревьевНаличие и тип охраны деревьев
5)user_type Категория пользователя, который собрал эту точку дерева
6)borocode Геокодирование на основе местоположения точки дерева
7)boroname Геокодирование на основе местоположения точки дерева

ВОПРОС 3. Столбец с макимальным количеством уникальных значений категориального признака?

In [21]:
ctgrl_priz = ['status','health','steward','guards','user_type','borocode','boroname']

for column in data.columns:
    if column in ctgrl_priz:
        print(column, data[column].nunique())

status 3
health 3
steward 4
guards 4
user_type 3
borocode 5


Максимальное число категориальных признаков у столбца borocode

ВОПРОС 4. Есть ли бинарные признаки?

1)curb_loc 
2)sidewalk 
3)root_stone 
4)root_grate 
5)root_other 
6)trunk_wire 
7)trnk_light 
8)trnk_other 
9)brch_light 
10)brch_shoe 
11)brch_other

ВОПРОС 5. Какие числовые признаки?

1)tree_id 
2)block_id 
3)created_at
4)tree_dbh 
5)stump_diam 
6)zipcode 
7)cb_num 
8)cncldist 
9)st_assem 
10)st_senate 
11)latitude 
12)longitude 
13)x_sp 
14)y_sp 

ВОПРОС 6. Есть ли пропуски?

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 683788 entries, 0 to 683787
Data columns (total 45 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   tree_id           683788 non-null  int64  
 1   block_id          683788 non-null  int64  
 2   created_at        683788 non-null  object 
 3   tree_dbh          683788 non-null  int64  
 4   stump_diam        683788 non-null  int64  
 5   curb_loc          683788 non-null  object 
 6   status            683788 non-null  object 
 7   health            652172 non-null  object 
 8   spc_latin         652169 non-null  object 
 9   spc_common        652169 non-null  object 
 10  steward           652173 non-null  object 
 11  guards            652172 non-null  object 
 12  sidewalk          652172 non-null  object 
 13  user_type         683788 non-null  object 
 14  problems          652124 non-null  object 
 15  root_stone        683788 non-null  object 
 16  root_grate        68

ВОПРОС 7. Сколько объектов с пропусками?

Основываясь на результате info() - видим, что у некоторых объектов есть пропуски:
1)health
2)spc_latin
3)spc_common
4)steward
5)guards
6)sidewalk
7)problems

*Данные признаки не обозначены в документации к датасету.
council district,census tract,bin,bbl              

ВОПРОС 8. Столбец с максимальным количеством пропусков?

In [23]:
data.isnull().sum(axis = 0)

tree_id                 0
block_id                0
created_at              0
tree_dbh                0
stump_diam              0
curb_loc                0
status                  0
health              31616
spc_latin           31619
spc_common          31619
steward             31615
guards              31616
sidewalk            31616
user_type               0
problems            31664
root_stone              0
root_grate              0
root_other              0
trunk_wire              0
trnk_light              0
trnk_other              0
brch_light              0
brch_shoe               0
brch_other              0
address                 0
postcode                0
zip_city                0
community board         0
borocode                0
borough                 0
cncldist                0
st_assem                0
st_senate               0
nta                     0
nta_name                0
boro_ct                 0
state                   0
latitude                0
longitude   

Столбец problems имеет 31664, но не имеет документирования. Следовательно, столбец/столбцы с наибольшим/им числом/ами пропусков - это spc_latin = 31619 и spc_common = 31619

ВОПРОС 9. Есть ли на ваш взгляд выбросы, аномальные значения?

In [24]:
for column in data.columns:
    print(data[column].value_counts())

2047      1
638877    1
626587    1
624538    1
630681    1
         ..
298436    1
308675    1
310722    1
304577    1
2049      1
Name: tree_id, Length: 683788, dtype: int64
204850    450
602362    358
208115    250
506756    206
233208    197
         ... 
341732      1
204966      1
202201      1
308948      1
316900      1
Name: block_id, Length: 101390, dtype: int64
2015-10-27T00:00:00.000    6852
2015-10-13T00:00:00.000    6676
2015-10-26T00:00:00.000    6258
2015-10-29T00:00:00.000    6223
2015-10-15T00:00:00.000    6085
                           ... 
2016-10-03T00:00:00.000       9
2016-10-04T00:00:00.000       7
2016-03-02T00:00:00.000       6
2016-04-03T00:00:00.000       4
2016-03-14T00:00:00.000       2
Name: created_at, Length: 483, dtype: int64
4      60372
3      54454
2      41977
5      41642
11     37978
       ...  
75         1
151        1
154        1
159        1
450        1
Name: tree_dbh, Length: 146, dtype: int64
0      666134
4         966
5         939
3 

Аномальных значений и выбросов не замечено.

ВОПРОС 10. Столбец с максимальным средним значением после нормировки признаков через стандартное отклонение?

Нормализация через стандартное отклонение:

In [25]:
import math
arr = ['tree_id','block_id','tree_dbh','stump_diam','postcode','community board','cncldist','st_assem','st_senate','latitude','longitude','x_sp','y_sp']
for column in arr:
    data[column] = (data[column] - data[column].mean())/(math.sqrt(data[column].var()))

Поиск максимального среднего

In [26]:
avrg = -math.inf
for column in arr:
    if data[column].mean() > avrg:
        avrg = data[column].mean()
print(avrg, column)

1.3488429021078534e-11 y_sp


Столбец с максимальным средним значением после нормирования - это y_sp

ВОПРОС 11. Столбец с целевым признаком?

Так как, мне надо выделить районы, в которых состояние деревьев аномально хорошее или плохое; 
выделить наилучший район по состоянию деревьев; 
выделить виды деревьев (или более общие элементы классификации), обладающие наилучшими показателями «здоровья», то будем оценивать столбцы:

1)status - Указывает, является ли дерево живым, стоящим мертвым или пнем.

2)zip_city - Город, полученный из почтового индекса

3)root_stone - Проблемы с корнями, вызванные брусчаткой в ложе дерева 

4)root_grate - Проблемы с корнями, вызванные металлическими решетками 

5)root_other - Наличие других проблем с корнями 

6)trunk_wire - Проблемы со стволом, вызванные тросом или проводами 

7)trnk_light - Проблемы с багажником, вызванные освещением 

8)trnk_other - Наличие других проблем со стволом 

9)brch_light - Проблемы с ветками, вызванные освещением или проводами

10)brch_shoe - Проблемы с ветвями, вызванные обувью 

11)brch_other - Наличие других проблем с ветвями 

12)spc_common - Общее название породы деревьев

ВОПРОС 12. Сколько объектов попадает в тренировочную выборку при использовании train_test_split с параметрами test_size = 0.3, random_state = 42?

Исключаем столбец с целевыми признаками из датасета для тренировки.

In [27]:
data_training = data.drop(['status','zip_city','root_stone','root_grate','root_other','trunk_wire','trnk_light','trnk_other','brch_light','brch_shoe','brch_other','spc_common'], axis = 1)
data_target = data[['status','zip_city','root_stone','root_grate','root_other','trunk_wire','trnk_light','trnk_other','brch_light','brch_shoe','brch_other','spc_common']]

In [28]:
data_training.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 683788 entries, 0 to 683787
Data columns (total 33 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   tree_id           683788 non-null  float64
 1   block_id          683788 non-null  float64
 2   created_at        683788 non-null  object 
 3   tree_dbh          683788 non-null  float64
 4   stump_diam        683788 non-null  float64
 5   curb_loc          683788 non-null  object 
 6   health            652172 non-null  object 
 7   spc_latin         652169 non-null  object 
 8   steward           652173 non-null  object 
 9   guards            652172 non-null  object 
 10  sidewalk          652172 non-null  object 
 11  user_type         683788 non-null  object 
 12  problems          652124 non-null  object 
 13  address           683788 non-null  object 
 14  postcode          683788 non-null  float64
 15  community board   683788 non-null  float64
 16  borocode          68

Разделим данные на тренировочные и тестовые

In [29]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(data_training, data_target, test_size=.3, random_state=42)

In [30]:
x_train

tree_id  block_id               created_at  tree_dbh  stump_diam  \
160056 -0.005521  1.727696  2015-10-22T00:00:00.000  2.260704   -0.131438   
287015  0.606543  0.820626  2015-12-02T00:00:00.000 -0.146713   -0.131438   
113753 -0.383871  0.857182  2015-10-05T00:00:00.000 -1.063825   -0.131438   
423584  1.349352 -0.044332  2016-07-29T00:00:00.000 -1.293102    5.339286   
238843  0.321215  0.812397  2015-11-05T00:00:00.000 -0.261352   -0.131438   
...          ...       ...                      ...       ...         ...   
259178  0.369759  1.711055  2015-11-09T00:00:00.000  0.655759   -0.131438   
365838  1.000615 -0.092243  2016-06-02T00:00:00.000 -0.719908   -0.131438   
131932 -0.123999  1.738215  2015-10-16T00:00:00.000 -0.834547   -0.131438   
671155 -0.859664  1.670677  2015-08-29T00:00:00.000  0.197203   -0.131438   
121958 -0.297998 -1.795802  2015-10-09T00:00:00.000  0.197203   -0.131438   

       curb_loc health          spc_latin steward   guards  ...  boro_ct  \
160056   OnCurb   Good  Quercus palustris    None     None  ...  2031000   
287015   OnCurb   Good        Acer rubrum    None     None  ...  5018702   
113753   OnCurb   Poor      Ginkgo biloba    None     None  ...  5022600   
423584   OnCurb    NaN                NaN     NaN      NaN  ...  4000800   
238843   OnCurb   Fair   Pyrus calleryana    None  Helpful  ...  5007400   
...         ...    ...                ...     ...      ...  ...      ...   
259178   OnCurb   Good      Tilia cordata    None     None  ...  2020400   
365838   OnCurb   Good           Fraxinus    1or2   Unsure  ...  4024300   
131932   OnCurb   Good             Prunus    None     None  ...  2046202   
671155   OnCurb   Good      Quercus rubra    None     None  ...  2038900   
121958   OnCurb   Good   Pyrus calleryana    1or2     None  ...  1013600   

           state  latitude longitude      x_sp      y_sp  council district  \
160056  New York  1.737253  0.695695  0.693014  1.737568              13.0   
287015  New York -1.090468 -1.652555 -1.652426 -1.090721              49.0   
113753  New York -1.901406 -2.481750 -2.484449 -1.899898              51.0   
423584  New York -0.143172  0.537737  0.538447 -0.143219              32.0   
238843  New York -1.211493 -1.162457 -1.161957 -1.212322              50.0   
...          ...       ...       ...       ...       ...               ...   
259178  New York  1.545682  0.599417  0.597330  1.545810              18.0   
365838  New York  0.418480  0.204343  0.204545  0.417969              30.0   
131932  New York  1.875252  0.717991  0.714998  1.875619              12.0   
671155  New York  1.735965  0.322483  0.320883  1.735666              15.0   
121958  New York  0.780872 -0.205017 -0.204564  0.780006               5.0   

       census tract        bin           bbl  
160056        310.0  2050996.0  2.044010e+09  
287015      18702.0  5125942.0  5.008070e+09  
113753        226.0  5148059.0  5.074460e+09  
423584          8.0  4185430.0  4.089660e+09  
238843         74.0  5135775.0  5.031220e+09  
...             ...        ...           ...  
259178        204.0  2042176.0  2.039950e+09  
365838        243.0  4056075.0  4.024190e+09  
131932      46202.0  2112917.0  2.047960e+09  
671155        389.0  2094658.0  2.030780e+09  
121958        136.0  1051113.0  1.015778e+09  

[478651 rows x 33 columns]

В тренировочную выборку попадет 478651 объектов

ВОПРОС 13. Между какими признаками наблюдается линейная зависимость (корреляция)?

Преобразуем текстовые данные в числовые

In [ ]:
texted = ['created_at', 'curb_loc', 'health', 'spc_latin', 'steward', 'guards', 'sidewalk', 'user_type', 'problems', 'address', 'borough', 'nta', 'nta_name', 'state']
for column in texted:
    cur_set = set(data[column])
    i = 0
    for item in cur_set:
        data[column] = data[column].replace(item, i)
        i = i + 1

Проверим коэффициент VIF (Variable Inflation Factors) для всех признаков. Если VIF > 10, то есть четкая линейная зависимость.

In [13]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
vif_data = pd.DataFrame()
vif_data["column_name"] = data_training.columns
vif_data["VIF"] = [variance_inflation_factor(data_training.values, i) for i in range(len(data_training.columns))]

TypeError: '>=' not supported between instances of 'str' and 'float'